### ISSUE TITLE
Flakey behavior when writing to BigQuery

### ISSUE THREAD

**Issue by imrimt** [GoogleCloudDataproc/spark-bigquery-connector#1170]
This is a recreation of #1131 with more details:

Connector's version: `spark-bigquery-with-dependencies_2.12:0.32.2@jar` 
Dataproc Spark version:  3.1.3
Language: Java 8

We are using the connector to both read from a BigQuery table and write to another BigQuery table. We are seeing issues during write. We just ran into it again in our production instance, with essentially the same errors as I pasted in the previous ticket: (I used the old logs since it already sanitizes sensitive information).

We use the following code for read and write:

```
final Dataset<Row> reader = spark.read()
  .format("bigquery")
  .option("project", <project>)
  .option("parentProject", <project>)
  .option("viewsEnabled", true)
  .option("gcpAccessToken", <token>) 
  .load(<fully_qualified_table>);

data.write().format("bigquery")
  .mode(SaveMode.Overwrite)
  .option("project", <project>)
  .option("parentProject", <project>)
  .option("writeMethod", "direct")
  .option("gcpAccessToken", <token>)
  .save(<fully_qualified_table>);
```

The following logs are seen in the failing Dataproc executor:
```
23/11/20 16:24:28 INFO com.google.cloud.spark.bigquery.direct.BigQueryRDDFactory: Created read session for table '<project>.<dataset>.<source-table>': projects/<project>/locations/us-east1/sessions/CAISDHhVa1pSU3JXaWVidhoCdngaAnZs
23/11/20 16:24:29 ERROR org.apache.spark.scheduler.TaskSetManager: task 0.0 in stage 39.0 (TID 39) had a not serializable result: com.google.cloud.spark.bigquery.repackaged.io.grpc.Status
Serialization stack:
  - object not serializable (class: com.google.cloud.spark.bigquery.repackaged.io.grpc.Status, value: Status{code=NOT_FOUND, description=Requested entity was not found. Entity: projects/<project>/datasets/<dataset>/tables/<target-table>/streams/Cig2ZTE5NTc4OC0wMDAwLTIyN2MtOTQxNS1mNGY1ZTgwZjdiYjQ6czEx, cause=null})
  - writeObject data (class: java.lang.Throwable)
  - object (class com.google.cloud.spark.bigquery.repackaged.io.grpc.StatusRuntimeException, com.google.cloud.spark.bigquery.repackaged.io.grpc.StatusRuntimeException: NOT_FOUND: Requested entity was not found. Entity: projects/<project>/datasets/<dataset>/tables/<target-table>/streams/Cig2ZTE5NTc4OC0wMDAwLTIyN2MtOTQxNS1mNGY1ZTgwZjdiYjQ6czEx)
  - writeObject data (class: java.lang.Throwable)
  - object (class com.google.cloud.spark.bigquery.repackaged.com.google.api.gax.rpc.NotFoundException, com.google.cloud.spark.bigquery.repackaged.com.google.api.gax.rpc.NotFoundException: com.google.cloud.spark.bigquery.repackaged.io.grpc.StatusRuntimeException: NOT_FOUND: Requested entity was not found. Entity: projects/<project>/datasets/<dataset>/tables/<target-table>/streams/Cig2ZTE5NTc4OC0wMDAwLTIyN2MtOTQxNS1mNGY1ZTgwZjdiYjQ6czEx)
  - writeObject data (class: java.lang.Throwable)
  - object (class java.util.concurrent.ExecutionException, java.util.concurrent.ExecutionException: com.google.cloud.spark.bigquery.repackaged.com.google.api.gax.rpc.NotFoundException: com.google.cloud.spark.bigquery.repackaged.io.grpc.StatusRuntimeException: NOT_FOUND: Requested entity was not found. Entity: projects/<project>/datasets/<dataset>/tables/<target-table>/streams/Cig2ZTE5NTc4OC0wMDAwLTIyN2MtOTQxNS1mNGY1ZTgwZjdiYjQ6czEx)
  - writeObject data (class: java.lang.Throwable)
  - object (class com.google.cloud.bigquery.connector.common.BigQueryConnectorException, com.google.cloud.bigquery.connector.common.BigQueryConnectorException: Could not retrieve AppendRowsResponse)
  - field (class: com.google.cloud.spark.bigquery.write.DataSourceWriterContextPartitionHandler$1, name: val$e, type: class java.lang.Exception)
  - object (class com.google.cloud.spark.bigquery.write.DataSourceWriterContextPartitionHandler$1, com.google.cloud.spark.bigquery.write.DataSourceWriterContextPartitionHandler$1@11586cf4)
  - element of array (index: 0)
  - array (class [Ljava.lang.Object;, size 1); not retrying
23/11/20 16:24:29 WARN com.google.cloud.spark.bigquery.write.context.BigQueryDirectDataSourceWriterContext: BigQuery Data Source writer 264dc07d-49f8-4174-9596-82d9484f9202 aborted
23/11/20 16:24:29 ERROR com.tamr.flex.copy.spark.Errors: Copy failed. Exit.
com.google.cloud.bigquery.connector.common.BigQueryConnectorException: unexpected issue trying to save [string_field: string, array_field: array<string> ... 1 more field]
  at com.google.cloud.spark.bigquery.write.BigQueryDataSourceWriterInsertableRelation.insert(BigQueryDataSourceWriterInsertableRelation.java:128)
  at com.google.cloud.spark.bigquery.write.CreatableRelationProviderHelper.createRelation(CreatableRelationProviderHelper.java:54)
  at com.google.cloud.spark.bigquery.BigQueryRelationProvider.createRelation(BigQueryRelationProvider.scala:107)
  at org.apache.spark.sql.execution.datasources.SaveIntoDataSourceCommand.run(SaveIntoDataSourceCommand.scala:46)
  at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:70)
  at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:68)
  at org.apache.spark.sql.execution.command.ExecutedCommandExec.doExecute(commands.scala:90)
  at org.apache.spark.sql.execution.SparkPlan.$anonfun$execute$1(SparkPlan.scala:180)
  at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:218)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
  at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:215)
  at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:176)
  at org.apache.spark.sql.execution.QueryExecution.toRdd$lzycompute(QueryExecution.scala:133)
  at org.apache.spark.sql.execution.QueryExecution.toRdd(QueryExecution.scala:132)
  at org.apache.spark.sql.DataFrameWriter.$anonfun$runCommand$1(DataFrameWriter.scala:989)
  at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$5(SQLExecution.scala:103)
  at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:163)
  at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:90)
  at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:775)
  at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
  at org.apache.spark.sql.DataFrameWriter.runCommand(DataFrameWriter.scala:989)
  at org.apache.spark.sql.DataFrameWriter.saveToV1Source(DataFrameWriter.scala:438)
  at org.apache.spark.sql.DataFrameWriter.saveInternal(DataFrameWriter.scala:415)
  at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:293)
  <omitted internal details>
```

In short, I don't think we are doing anything custom with the connector. And subsequent retries do work, but the issues seem to still be coming back every now and then.

**Comment by vishalkarve15** [comment]
We don't have logs older than 30 days. Can you please provide a more recent instance of this?
@agrawal-siddharth

**Comment by imrimt** [comment]
[masked_logs.txt](https://github.com/GoogleCloudDataproc/spark-bigquery-connector/files/13982009/masked_logs.txt)
Here's a more current log that happened on our production instance 4 days ago. The target region is `us-east-1` and project ID is `prdtrellixba9df138c0d64aa8ab22`. I had to mask the dataset and table name for security purposes.

**Comment by vishalkarve15** [comment]
@imrimt we did not find anything abnormal about this read stream in our logs. I'd suggest you to open a request with Google Support so that you can share more specific details more comfortably and effectively.
